File: practice/easy/sums_in_a_triangle.pl
#
# Problem: Sums in a triangle
# URL: http://www.codechef.com/problems/SUMTRIAN
#
# Author: Ronald Kaiser
# Email: raios dot catodicos at gmail dot com
#
# Observation: This is a so much better version than the
# previous one, but it can be better.
# This code didn't pass in the Time limit of codechef, although
# the same algorithm in C code, runs pretty well. 
# Investigate a better approach in Perl.
#


chomp(my $n = <>);

my @triangle = [];
while(<>) {
    chomp($_);
    for(1 .. $_) {
        my @line = split(/\s+/, <>);
        @triangle[$_ - 1] = \@line;
    }  
    for (my $i = $_-2; $i >= 0; $i--) { 
        for (my $j = 0; $j <= $i; $j++) {
            $triangle[$i][$j] += ($triangle[$i+1][$j] > $triangle[$i+1][$j+1]) ? 
                                    $triangle[$i+1][$j] : $triangle[$i+1][$j+1];
        }
    };
    print "$triangle[0]->[0]" . "\n";

    $n--;
    last if $n == 0;
};

File: practice/easy/sums_in_a_triangle.c
/*
 Problem: Sums in a triangle
 URL: http://www.codechef.com/problems/SUMTRIAN
 
 Author: Ronald Kaiser
 Email: raios dot catodicos at gmail dot com

 Observation: Used memoization technique. It can be a
 little bit faster. Actually, it's not necessary to allocate
 all the matrix, just the larger line of it -- last one.
 But it solves the problem in codechef within the given 
 Time Limit.
 Brute force, without memoization gives a O(2^n).
 With memoization: O(n^2).
*/


#include<stdio.h>

int m[101][101];

int main() {
    int n, t, i, j, k, lines;
    
    scanf("%d\n", &n);
    for(t = 0; t < n; t++) {
        scanf("%d\n", &lines);
    
        for(i = 0; i < lines; i++) 
            for(k = 0; k <= i; k++) 
               scanf("%d", &m[i][k]);

        for(i = lines - 2; i >= 0; i--)
            for(j = 0; j <= i; j++)
                m[i][j] += m[i+1][j] > m[i+1][j+1] ? m[i+1][j] : m[i+1][j+1];

         printf("%d\n", m[0][0]);
    }
    return 0;
}
